pkg/agent/core/metrics.go

```diff
@@ -4,78 +4,143 @@ package core
 
 import (
 	"fmt"
-	"strconv"
-	"strings"
+	"io"
+
+	promtypes "github.com/prometheus/client_model/go"
+	promfmt "github.com/prometheus/common/expfmt"
+	"github.com/tychoish/fun/erc"
 
 	"github.com/neondatabase/autoscaling/pkg/api"
 )
 
-type Metrics struct {
-	LoadAverage1Min  float32
-	MemoryUsageBytes float32
+type SystemMetrics struct {
+	LoadAverage1Min  float64
+	MemoryUsageBytes float64
 }
 
-func (m Metrics) ToAPI() api.Metrics {
+func (m SystemMetrics) ToAPI() api.Metrics {
 	return api.Metrics{
-		LoadAverage1Min:  m.LoadAverage1Min,
+		LoadAverage1Min:  float32(m.LoadAverage1Min),
 		LoadAverage5Min:  nil,
 		MemoryUsageBytes: nil,
 	}
 }
 
-// ReadMetrics generates Metrics from vector.dev's host metrics output, or returns error on failure
+type LFCMetrics struct {
+	CacheHitsTotal   float64
+	CacheMissesTotal float64
+	CacheWritesTotal float64
+
+	ApproximateWorkingSetSizeTotal float64 // approximate_working_set_size
+}
+
+// FromPrometheus represents metric types that can be parsed from prometheus output.
 //
-// This function could be more efficient, but realistically it doesn't matter. The size of the
-// output from node_exporter/vector is so small anyways.
-func ReadMetrics(nodeExporterOutput []byte, loadPrefix string) (m Metrics, err error) {
-	lines := strings.Split(string(nodeExporterOutput), "\n")
-
-	getField := func(linePrefix, dontMatch string) (float32, error) {
-		var line string
-		for _, l := range lines {
-			if strings.HasPrefix(l, linePrefix) && (len(dontMatch) == 0 || !strings.HasPrefix(l, dontMatch)) {
-				line = l
-				break
-			}
-		}
-		if line == "" {
-			return 0, fmt.Errorf("No line in metrics output starting with %q", linePrefix)
-		}
+// It is parameterized by the config 'C' passed to the parsing function, to allow things like metric
+// names to be defined externally, rather than compile-time constants.
+type FromPrometheus interface {
```

Comment:
WDYT about naming it `Metrics`, and using as a type parameter constraint instead of any? You mention "metric types" in the comment anyway, and this would help readability.

Also, two bottom lines are no longer true.

Reply:
> Also, two bottom lines are no longer true.

Ah good catch, fixed in 662613d170363fa9cf240e5a0225e0c2f1c09674.

> and using as a type parameter constraint instead of any

Not quite sure I follow, could you elaborate?

Reply:
Hm, that's fair. In rust, it's generally an anti-pattern to restrict generic types more than they need to be, but I'm not sure if it's the same in Go.

I like the idea of using it to signal purpose. Changed from `M any` to `M FromPrometheus` in d0991a9d2447ff1c2a37e0cf2241422df51dea52